### PR TITLE
Handle token binding validation property migration

### PR DESCRIPTION
**Pull request by Buddhimah**

### CODE REVIEW

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/bean/OIDCSPInfo.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+package org.wso2.carbon.is.migration.service.v5110.bean;
+
+/**
+ * OIDC Service Provider Info.
+ */
+public class OIDCSPInfo {
+
+    private String consumerKey;
+    private int tenantID;
+
```
Comment (ayshsandu) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/bean/OIDCSPInfo.java:25]: Add a constructor to set consumerKey and tenantID, and refer it wherever the setters are called.

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/dao/TokenBindingDAO.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            "CONSUMER_KEY = ? AND TENANT_ID = ? AND PROPERTY_KEY = ?";
+    public static final String ADD_TOKEN_BINDING_VALIDATION_PROPERTY = "INSERT INTO IDN_OIDC_PROPERTY (TENANT_ID, " +
+            "CONSUMER_KEY, PROPERTY_KEY, PROPERTY_VALUE) VALUES (?, ?, ?, ?)";
+    public static final String TENANT_ID = "TENANT_ID";
+    public static final String CONSUMER_KEY = "CONSUMER_KEY";
+    public static final String TOKEN_BINDING_TYPE = "tokenBindingType";
+    public static final String COOKIE = "cookie";
+    public static final String TOKEN_BINDING_VALIDATION = "tokenBindingValidation";
+
+    private static TokenBindingDAO instance = new TokenBindingDAO();
```
Comment (ayshsandu) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/dao/TokenBindingDAO.java:50]: Why does this class need to be singleton?

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/dao/TokenBindingDAO.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                while (resultSet.next()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Method to check whether token binding validation property exists.
```
Comment (ayshsandu) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/dao/TokenBindingDAO.java:108]: Change the method description accordingly

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/migrator/TokenBindingValidationMigrator.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            reportUtil.writeMessage(
+                    String.format("%20s | %20s ", "Tenant ID", "Consumer Key"));
+
+            log.info(Constant.MIGRATION_LOG + "Started the dry run token binding validation migration.");
+            List<OIDCSPInfo> oidcSpInfoList;
+            try (Connection connection = getDataSource(Schema.IDENTITY.getName()).getConnection()) {
+                try {
+                    oidcSpInfoList = TokenBindingDAO.getInstance().getOIDCServiceProvidersData(connection);
+                    for (OIDCSPInfo oidcspInfo : oidcSpInfoList) {
+                        if (!TokenBindingDAO.getInstance().isTokenBindingValidationPropertyExists(connection,
```
Comment (ayshsandu) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/migrator/TokenBindingValidationMigrator.java:67]: Refer previous comment on TokenBindingDAO.getInstance(). If not needed, change instance initiation accordingly.

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/bean/OIDCSPInfo.java (hunk 1 above)
Comment (Buddhimah) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/bean/OIDCSPInfo.java:25]: Done

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/dao/TokenBindingDAO.java (hunk 2 above)
Comment (Buddhimah) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/dao/TokenBindingDAO.java]: Changed

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/dao/TokenBindingDAO.java (hunk 3 above)
Comment (Buddhimah) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/dao/TokenBindingDAO.java]: Done

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/migrator/TokenBindingValidationMigrator.java (hunk 4 above)
Comment (Buddhimah) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/migrator/TokenBindingValidationMigrator.java]: Done

File: components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/dao/TokenBindingDAO.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * TokenBindingDAO implementation.
+ */
+public class TokenBindingDAO {
+
+    public static final String RETRIEVE_OIDC_SERVICE_PROVIDERS =
```
Comment (omindu) [inline components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/dao/TokenBindingDAO.java:37]: Can there be OIDC SPs without an entry in IDN_OIDC_PROPERTY table ?

Comment (Buddhimah) [reply components/org.wso2.is.migration/migration-service/src/main/java/org/wso2/carbon/is/migration/service/v5110/dao/TokenBindingDAO.java:37]: https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java#L155

I have checked the code when we add an OAuthApp we call the addServiceProviderOIDCProperties method to Add properties.

**Review by omindu — APPROVED** [review]

**Comment by Buddhimah** [comment]
There are two properties that need to be there to happen token binding validation

1) tokenBindingType
2) tokenBindingValidation - enabled

Token Binding was a feature introduced to 5.10 GA. In 5.10 GA there was no property such as tokenBindingValidation hence by default for all SP's token binding validation will happen. After that, we have provided a WUM for 5.10 with a property tokenBindingValidation which gives the capability to enable or disable token binding validation for SP's. But since the already existing SP's had token Binding validation enabled, If the tokenBindingValidation property does not exist for a sp (SP created before WUM) in the code level we will consider it as a token binding validation enabled sp. But in 5.11.0 We have done it the correct way by default  If the tokenBindingValidation property does not exist for a sp we will consider it as a token binding validation disabled sp. [1]

Let's assume an SP is migrating from 5.10 to 5.11 there are few scenarios

1) Token Binding Validation Property Exists and it's true - Will keep it that way 
2) Token Binding Validation Property Exists and it's false - Will keep it that way 
3) Token Binding Validation Property Does not exist and Token Binding type is Cookie - Add a property as true by migration client
4) Token Binding Validation Property Does not exist and Token Binding type is Session - Will keep it that way

Let's assume that SP is migrating from 5.9.0 to 5.11

1) In 5.9.0 Token Binding Validation will be not available as a feature. So both above properties will not be there. Hence after migration also TokenBindingValidation Will Not Happen 

Let's assume that SP is migrating from 5.4.0 to 5.11

Before 5.5 this table IDN_OIDC_PROPERTY was not there So non of these properties will be there. In that case also both these properties will be not there hence after migration also token binding validation will not happen. 


[1].https://github.com/wso2-support/identity-inbound-auth-oauth/pull/732#issuecomment-698795757

**Review by ayshsandu — APPROVED** [review]

**Pull request merged**